File: examples/gasless-transfer-7702/go/main.go
package main

import (
	"encoding/hex"
	"fmt"
	"os"

	"github.com/zerodevapp/zerodev-omni-sdk/bindings/go/aa"
)

func main() {
	fmt.Println("===========================================")
	fmt.Println("  ZeroDev Gasless Transfer (EIP-7702) — Go")
	fmt.Println("===========================================")
	fmt.Println()

	// ── Step 1: Read environment variables ──────────────────────────
	projectID := os.Getenv("ZERODEV_PROJECT_ID")
	if projectID == "" {
		fmt.Fprintln(os.Stderr, "Error: ZERODEV_PROJECT_ID environment variable is required.")
		fmt.Fprintln(os.Stderr, "Usage:")
		fmt.Fprintln(os.Stderr, "  export ZERODEV_PROJECT_ID=<your-project-id>")
		fmt.Fprintln(os.Stderr, "  go run main.go")
		os.Exit(1)
	}

	fmt.Printf("[1/6] Configuration loaded (project: %s, chain: Sepolia 11155111)\n", projectID)

	// ── Step 2: Create context with ZeroDev gas + paymaster on Sepolia ─
	chainID := uint64(11155111) // Sepolia

	ctx, err := aa.NewContext(projectID, "", "", chainID, aa.GasZeroDev, aa.PaymasterZeroDev)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating context: %v\n", err)
		os.Exit(1)
	}
	defer ctx.Close()

	fmt.Println("[2/6] Context created (gas: ZeroDev, paymaster: ZeroDev)")

	// ── Step 3: Generate a fresh EOA key ───────────────────────────
	//
	// EIP-7702 delegation lets *any* EOA act as a Kernel smart account — no
	// CREATE2 deployment is needed. We generate a key to demonstrate the
	// delegation-install path; replace with aa.LocalSigner(pk) to reuse an
	// existing EOA.
	signer, err := aa.GenerateSigner()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error generating signer: %v\n", err)
		os.Exit(1)
	}
	defer signer.Close()

	fmt.Println("[3/6] Signer created (random EOA key generated)")

	// ── Step 4: Create the 7702 account (sender == EOA) ────────────
	//
	// Unlike NewAccount, this does NOT take an index — a 7702 account's
	// address is the EOA itself. On the first UserOperation the SDK signs an
	// authorization tuple and attaches it via `eip7702Auth`; subsequent UserOps
	// skip the auth once delegation is installed on-chain.
	account, err := ctx.NewAccount7702(signer, aa.KernelV3_3)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating 7702 account: %v\n", err)
		os.Exit(1)
	}
	defer account.Close()

	addrHex, err := account.GetAddressHex()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting account address: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("[4/6] Smart account (EOA + delegation) ready: %s\n", addrHex)

	// ── Step 5: Build a call (send 0 ETH to self — noop) ───────────
	addr, err := account.GetAddress()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting account address bytes: %v\n", err)
		os.Exit(1)
	}

	calls := []aa.Call{
		{
			Target:   addr,
			Value:    [32]byte{}, // 0 ETH
			Calldata: []byte{},   // empty calldata
		},
	}

	fmt.Println("[5/6] Sending sponsored UserOp with eip7702Auth...")

	// ── Step 6: Send the UserOp (build + sponsor + sign + submit) ──
	hash, err := account.SendUserOp(calls)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error sending UserOp: %v\n", err)
		os.Exit(1)
	}

	hashHex := "0x" + hex.EncodeToString(hash[:])
	fmt.Printf("       UserOp hash: %s\n", hashHex)

	// ── Step 7: Wait for the receipt ────────────────────────────────
	fmt.Println("[6/6] Waiting for receipt (timeout: 60s, poll: 2s)...")

	receipt, err := account.WaitForUserOperationReceipt(hash, 60000, 2000)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error waiting for receipt: %v\n", err)
		os.Exit(1)
	}

	// ── Print results ───────────────────────────────────────────────
	fmt.Println()
	fmt.Println("===========================================")
	fmt.Println("  UserOp Receipt")
	fmt.Println("===========================================")
	fmt.Printf("  Success:      %v\n", receipt.Success)

	if txHash, ok := receipt.Receipt["transactionHash"].(string); ok {
		fmt.Printf("  Tx Hash:      %s\n", txHash)
	}

	fmt.Printf("  Gas Used:     %s\n", receipt.ActualGasUsed)
	fmt.Printf("  Gas Cost:     %s\n", receipt.ActualGasCost)
	fmt.Printf("  Sender (EOA): %s\n", receipt.Sender)
	fmt.Printf("  Paymaster:    %s\n", receipt.Paymaster)
	fmt.Printf("  UserOp Hash:  %s\n", receipt.UserOpHash)
	fmt.Printf("  Entry Point:  %s\n", receipt.EntryPoint)
	fmt.Println("===========================================")

	if !receipt.Success {
		fmt.Fprintf(os.Stderr, "\nUserOp execution reverted: %s\n", receipt.Reason)
		os.Exit(1)
	}

	fmt.Println("\n7702 gasless transfer completed successfully!")
}

File: examples/gasless-transfer-7702/kotlin/src/main/kotlin/Main.kt
import dev.zerodev.aa.*

// ---------------------------------------------------------------------------
// EIP-7702 Gasless Transfer Example (Sepolia)
//
// Demonstrates the 7702 flow: a fresh EOA is generated, then we call
// `newAccount7702` so the account address IS the EOA (no CREATE2, no init
// code). The first UserOp signs an authorization tuple delegating the EOA
// to the Kernel v3.3 implementation; subsequent ops skip re-authorization.
//
// The paymaster covers gas, so the freshly generated EOA needs zero balance.
//
// Required environment variables:
//   ZERODEV_PROJECT_ID  — your ZeroDev project ID
// ---------------------------------------------------------------------------

fun main() {
    // 1. Read environment variables
    val projectId = System.getenv("ZERODEV_PROJECT_ID")
    if (projectId.isNullOrEmpty()) {
        System.err.println("Error: ZERODEV_PROJECT_ID environment variable is not set.")
        System.err.println("Usage: ZERODEV_PROJECT_ID=<id> ./gradlew run")
        System.exit(1)
    }

    val chainId = 11155111L // Sepolia

    // 2. Create a context with ZeroDev gas + paymaster middleware on Sepolia
    println("Creating context (chain: Sepolia $chainId)...")
    Context.create(
        projectId = projectId,
        chainId = chainId,
        gasMiddleware = GasMiddleware.ZERODEV,
        paymasterMiddleware = PaymasterMiddleware.ZERODEV,
    ).use { ctx ->
        println("Context created.")

        // 3. Generate a fresh EOA. For EIP-7702 the smart account address
        //    IS the EOA address — no CREATE2 derivation, so this EOA does
        //    not need any pre-funding.
        Signer.generate().use { signer ->
            println("Signer generated (fresh EOA).")

            // 4. Create a Kernel v3.3 7702 account. Today only V3_3 is
            //    supported for 7702.
            ctx.newAccount7702(signer, KernelVersion.V3_3).use { account ->
                println("7702 account created.")

                // 5. Account address == EOA address (assert it for clarity).
                val address = account.getAddress()
                println("Smart account address (= EOA): $address")

                // 6. Build a call: send 0 ETH to self (gasless noop). On the
                //    first UserOp the SDK attaches an eip7702Auth tuple so
                //    the EOA is delegated to Kernel v3.3.
                val calls = listOf(
                    Call(target = address, value = ByteArray(32), calldata = ByteArray(0)),
                )
                println("Sending 7702 gasless UserOp (0 ETH to self)...")

                // 7. Send the UserOp through the bundler
                val useropHash = account.sendUserOp(calls)
                println("UserOp submitted!")
                println("  UserOp hash: $useropHash")

                // 8. Wait for the UserOp to be included on-chain
                println("Waiting for on-chain receipt...")
                val receipt = account.waitForUserOperationReceipt(useropHash)

                // 9. Print receipt details
                println()
                println("=== 7702 UserOp Receipt ===")
                println("  Success:         ${receipt.success}")
                println("  UserOp hash:     ${receipt.userOpHash}")
                println("  Sender:          ${receipt.sender}")
                println("  Nonce:           ${receipt.nonce}")
                println("  Actual gas used: ${receipt.actualGasUsed}")
                println("  Actual gas cost: ${receipt.actualGasCost}")
                if (!receipt.paymaster.isNullOrEmpty()) {
                    println("  Paymaster:       ${receipt.paymaster}")
                }
                println("===========================")

                if (receipt.success) {
                    println("\nEIP-7702 gasless transfer completed successfully!")
                } else {
                    System.err.println("\nUserOp execution reverted.")
                    if (!receipt.reason.isNullOrEmpty()) {
                        System.err.println("Reason: ${receipt.reason}")
                    }
                    System.exit(1)
                }
            }
        }
    }
}
